### ISSUE TITLE
Too many kappa_ind_comp errors

### ISSUE THREAD

**Issue by jhmatthews** [sirocco-rt/sirocco#91]
When running a standard fiducial model with the current commit of dev (https://github.com/agnwinds/python/commit/ea9854de9a24d6ac9f6d173673a143e50537f6f6) I'm getting so many errors that the run quits.

Two problems to flag up here, and the reason they now cause me to quit is because I got rid of Warnings in favour of Errors.

The errors are as follows

```
kappa_ind_comp: frequency of photon (%e) is outside frequency range (%e - %e) of spectral model in cell %i band %i\n

kappa_ind_comp: no model exists in cell %i band %i\n
```

The second one is probably not one we want to flag as an error every time there is a photon with no model, as we already flag this in spectral estimators, and it also happens a lot in bands with low/no photon numbers.

I would like to understand why the top error occurs, and decide if we need to fix something.  Secondly, we should revisit the topic of warnings briefly. The problem is that you don't really want to be notified of this via a log statement as there would be way too much output, but also an Error is perhaps not appropriate.

Fixing this in at least some capacity is urgent because we can't currently run the dev version on a fiducial agn model. 

Nick provides this comment (thanks Nick)

"We have a spectral model, but it doesnt apply to the frequency in question. 
 clearly this is a slightly odd situation, where last time we didnt get a photon 
of this frequency, but this time we did. Still this should only happen in very 
sparse cells, so induced compton is unlikely to be important in such cells. 
We generate a warning, just so we can see if this is happening a lot"

Here's the relevant section of code

``` C
if (geo.xfreq[i] < freq && freq <= geo.xfreq[i + 1])    //We have found the correct model band
        {
          if (xplasma->spec_mod_type[i] > 0)    //Only bother if we have a model in this band
        {
            if (freq > xplasma->fmin_mod[i] && freq < xplasma->fmax_mod[i]) //The spectral model is defined for the frequency in question
            {
            if (xplasma->spec_mod_type[i] == SPEC_MOD_PL)   //Power law model
                {               
                J = pow(10,(xplasma->pl_log_w[i]+log10(freq)*xplasma->pl_alpha[i]));
                }
            else if (xplasma->spec_mod_type[i] == SPEC_MOD_EXP) //Exponential model
                {
                J = xplasma->exp_w[i] * exp ((-1 * H * freq) / (BOLTZMANN * xplasma->exp_temp[i]));
                }
            else
                {
                Error("kappa_ind_comp - unknown spectral model (%i) in band %i\n",xplasma->spec_mod_type[i], i);
                J = 0.0;    //Something has gone wrong
                }
            }
        else /* We have a spectral model, but it doesnt apply to the frequency in question. 
             clearly this is a slightly odd situation, where last time we didnt get a photon 
             of this frequency, but this time we did. Still this should only happen in very 
             sparse cells, so induced compton is unlikely to be important in such cells. 
             We generate a warning, just so we can see if this is happening a lot */
            {
            Error ("kappa_ind_comp: frequency of photon (%e) is outside frequency range (%e - %e) of spectral model in cell %i band %i\n",
                        freq,xplasma->fmin_mod[i],xplasma->fmax_mod[i],xplasma->nplasma,i); 
            //This is unlikely to happen very often, but if it does, we should probably know about it
            J = 0.0; //We 
            }
        }
         else /* There is no model in this band - this should not happen very often  */
        {   
        J = 0.0;    //THere is no modelin this band, so the best we can do is assume zero J
        Error ("kappa_ind_comp: no model exists in cell %i band %i\n",xplasma->nplasma,i); 
        //This is unlikely to happen very often, but if it does, we should probably know about it
        }
```

**Comment by Higginbottom** [comment]
Well, I'm not surprised,since this is _precisely_ the error which caused me to introduce the warning scheme. 
To comment on your comments:

The second error is _not_ already mentioned in spectral estimators. Spectral estimators returns an error if it cannot make a model based upon the photons which flew in the previous cycle. If, in the _next_ cycle, photons with frequency that is in the range of this band are detected in the cell, then the kappa_ind_comp error is generated. 

The first error is more of an issue. It is simple to see why it happens - there were not enough photons to properly characterise the specific mean intensity for all frequencies in a band, so we truncate the band that the model applies to. Next cycle, for whatever reason, lots of photons outside that band are seen. What do we do? Do we extrapolate the model - this can be very bad, if we have a seriously pathological model (i.e. one that looks like J=1e-300 _nu^+100) then one could end up with *huge_ induced compton heating. I made the decision to ignore them, and generate a warning so interested users could see what was going on and consider upping the photon count. If we dont like the idea of warnings, the question here is how to generate a model from tiny numbers of photons that one can apply to all frequencies in a band.

Nick

**Comment by kslong** [comment]
I don't like the idea of warnings precisely because of the fact that it doesn't make us decide whether something is important or not important. Instead the person who discovered the problem just papers over it as a warning that no one worries about. What we need to decide is if this warning is really an error and if it's an error we need to fix it. If it's not an  error then we should be happy with that but we should make a decision. There are many worries we have about various aspects of python and therefore selecting out one to record as a warning is just not useful.

Knox

**Comment by jhmatthews** [comment]
Ok, so we are sticking as is with Errors. Now understanding what's actually going on here, I don't think we need to throw an error, because this is actually something you expect to happen as you converge on a correct model for J in a cell.

So I'm happy to just comment out the error messages and add some comments explaining the situation. 

The only decision to be made is whether we should perhaps adopt a dilute BB as J if there's no model in the band in this specific case? Otherwise it can just be set to 0 and sorted out in the next ionization cycle.

Agreed?

**Comment by Higginbottom** [comment]
Ok, so I agree that its not an error, but I'm slightly uneasy about just ignoring it entirely. I feel it is something that _I_ want to know about. It does after all tell us something about whether we are flying enough photons through a model

Of course one thing that I see now (and stupidly missed before) is that these errors should now be called "mean_intensity" not "kappa_ind_comp", since this routine is now also used by two_level_atom.

Should we just use a dilute blackbody? Hmmmm, I'm not sure. I think it is best to set it to zero. If this problem is arising, then we can be pretty certain that there are very few photons, so a guess of zero for J_nu is probably not a bad one!

**Comment by jhmatthews** [comment]
I was just thinking about exactly that after I wrote my comment. I'm starting to now believe that at least some kind of warning is a good idea. Let's see what @kslong says, we just need to agree on a convention for what to use when. The point is, I think, that this can happen in a perfectly decent model, so Error is not really appropriate, but it could also be a symptom of a real problem.

For the moment I've commented them out but this is very temporary so I can test some regression suites I'm working on.

**Comment by kslong** [comment]
Okay, but how would you handle this problem? How would you identify when there's a problem when there's not? Report the statistics? Or do you have an error really generated the numbers exceed a million, or ten million. If you all can agree on how to make us pay attention at some point, I am happy. 

Knox

**Comment by jhmatthews** [comment]
When Nick implemented Warning he did report the statistics, so it doesn't just sweep under the carpet. 

I think the problem is really just that you can't do this automatically. It's not possible (or at least, very, very difficult) to write the code in such a way that it knows whether or not it's bad or not if it has no photons in the band. Sometimes it can be perfectly normal (A region behind a dense region with significant Lyman photoabsorption, e.g.) to have no photons in a band, but sometimes it can be a sign that you don't have enough photons in the model. 

I see no way to really tell this without some kind of human interpretation, and therefore Warnings seem as anything to me, BUT they should only be used if the situation in question can occur a number of times in a perfectly normal model.

On a side note, to stop this sort of thing happening again (i.e. breaking a run) I've finally now got a test which runs nightly on the dev branch and checks that the fiducial model and old CV model run, and emails me if they don't. I'll add more runs to this as I go. We test v Python 78 which has been fairly rigorously tested itself, and the plots can be viewed on the following site (the repository itself is private but I can grant access to you guys if you'd like):

http://jhmatthews.github.io/pyregress/

**Comment by jhmatthews** [comment]
Closed by #99
